src/main/java/io/jenkins/plugins/pipeline_elasticsearch_logs/write/fluentd/FluentdWriter.java

```diff
@@ -0,0 +1,268 @@
+        this.configChanged = true;
+    }
+
+    public int getBufferCapacity() {
+        return bufferCapacity;
+    }
+
+    @DataBoundSetter
+    public void setBufferCapacity(int bufferCapacity) {
+        if (bufferCapacity < 0)
+            bufferCapacity = DEFAULT_BUFFER_CAPACITY;
+        this.bufferCapacity = bufferCapacity;
+        this.configChanged = true;
+    }
+
+
+    @Extension
+    @Symbol("fluentd")
+    public static class DescriptorImpl extends ElasticSearchWriteAccessDescriptor {
+        @Override
+        public String getDisplayName() {
+            return "Fluentd Elasticsearch Writer";
+        }
+    }
+
+    /**
+     * Logs the given object to Fluentd. This method blocks if Fluentd is not reachable and
+     * throws an IOException after TIMEOUT_MILLIS.
+     *
+     * @param data The data to post
+     * @throws IOException if something failed permanently
+     */
+    @Override
+    public void push(Map<String, Object> data) throws IOException {
+        if(fluentd == null || configChanged) initFluentdLogger();
+        try {
+
+            long timeMarker = System.currentTimeMillis();
+            boolean sent = fluentd.log(tag, data);
+            if(!sent) {
+                //Errors are not reported explicitly. But the latest error should be the reason for (sent == false)
+                Exception lastError = errorHandler.getLastError(timeMarker);
+                throw new IOException("Could not send logs to fluentd", lastError);
+            }
+
+            // (sent == true) does not necessarily mean the data has already been sent to fluentd.
+            // To ensure data can be sent we flush the library buffer every time.
+            flushWithRetry();
+
+        } catch (InterruptedException e) {
+            throw new IOException(e);
+        }
+    }
+
+
+    private void flushWithRetry() throws InterruptedException, IOException {
+        long start = System.currentTimeMillis();
+        long nextSystemLog = 0;
+        boolean failed = true;
+        while (failed) {
+            long timeMarker = System.currentTimeMillis();
+            fluentd.flush();
+            // Unfortunately flush() does not fail obviously. That's why we get errors indirectly via our ErrorHandler.
+            Exception flushError = errorHandler.getLastError(timeMarker);
+            if (flushError == null) {
+                failed = false;
+            } else {
+                if (nextSystemLog <= System.currentTimeMillis()) {
+                    LOGGER.warning(format(
+                            "Could not flush fluentd library buffer since %s seconds. FluentLogger (%s) %s. Configured host=%s, port=%s.",
+                            ((System.currentTimeMillis() - start) / 1000), fluentd.hashCode(),
+                            (fluentd.isConnected() ? "connected" : "not connected"), host, port));
+                    nextSystemLog = System.currentTimeMillis() + SYS_LOG_FREQUENCY;
```

Comment:
Why do we need `nextSystemLog` and `SYS_LOG_FREQUENCY`? (Only) to prevent the the system log is flooded with the same error messages?

Reply:
indeed probably not necessary. There will be only a log entry once for every retry cycle